aopy/data/bmi3d.py

```diff
@@ -564,4 +566,127 @@ def get_source_files(preproc_dir, subject, te_id, date):
+    trial start and end codes. 
+
+    Args:
+        preproc_dir (str): base directory where the files live
+        subjects (list of str): Subject name for each recording
+        ids (list of int): Block number of Task entry object for each recording
+        dates (list of str): Date for each recording
+        trial_start_codes (list): list of numeric codes representing the start of a trial
+        trial_end_codes (list): list of numeric codes representing the end of a trial
+        target_codes (list): ordered list of numeric codes representing the possible target indices
+        reward_codes (list): list of numeric codes representing rewards
+        penalty_codes (list): list of numeric codes representing penalties
+        df (DataFrame, optional): pandas DataFrame object to append. Defaults to None.
+        include_handdata (bool, optional): If True, includes hand trajectories in addition to cursor
+            trajectories. Defaults to False.
+        include_eyedata (bool, optional): If True, includes eye trajectories in addition to cursor
+            trajectories. Defaults to False.
+
+    Returns:
+        pd.DataFrame: pandas DataFrame containing the concatenated trial data
+    '''
+    if df is None:
+        df = pd.DataFrame()
+
+    entries = list(zip(subjects, dates, ids))
+    for subject, date, te in tqdm(entries): 
+
+        # Load data from bmi3d hdf 
+        try:
+            data, metadata = load_preproc_exp_data(preproc_dir, subject, te, date)
+        except:
+            print(f"Entry {subject} {date} {te} could not be loaded.")
+            traceback.print_exc()
+            continue
+        event_codes = data['events']['code']
+        event_times = data['events']['timestamp']
+
+        # Trial aligned event codes and event times
+        tr_seg, tr_t = get_trial_segments_and_times(event_codes, event_times, trial_start_codes, trial_end_codes)
+
+        # Get data segments 
+        cursor_traj = get_kinematic_segments(preproc_dir, subject, te, date, trial_start_codes, trial_end_codes, 
+                                                         datatype='cursor')[0].tolist()
+        hand_traj = [None] * len(tr_seg)
+        if include_handdata:
+            hand_traj = get_kinematic_segments(preproc_dir, subject, te, date, trial_start_codes, trial_end_codes, 
+                                                  datatype='hand')[0].tolist()
+        eye_traj = [None] * len(tr_seg)
+        if include_eyedata:
+            eye_traj = get_kinematic_segments(preproc_dir, subject, te, date, trial_start_codes, trial_end_codes, 
+                                                 datatype='eye')[0].tolist()
+
+        target_idx = [code[np.isin(code, target_codes)][0] - target_codes[0] if np.sum(np.isin(code, target_codes)) == 1 else 0 for code in tr_seg]
+        target_location = get_target_locations(preproc_dir, subject, te, date, target_idx).tolist()
```

Comment:
What is the point of including this?

Reply:
Because target idx can (and does) change across recordings, so if we don’t add it here it’s a pain to go through each row to add it later. But I do want this function to be useful for other kinds of trials, eg tracking. Not sure how this would work with those.

Reply:
I agree with Leo, target_locations are useful to calculate movement error metrics. Target locations changes with task setup parameters like target size, distance, etc. It is useful to keep this in the data frame and it is a pain to add later for each row. 
 
Also, does it make sense to have task specific data processing functions? example: for 2D center out vs tracking task could have different functions, but this function should work for center out task irrespective of the number of targets. 
would be useful to get Katherine's input if you are thinking about generalizing for tracking task as well. But make this a future update. Let's get this function in copy now.